### PR TITLE
MM-20580: Migrate 'components/admin_console/team_channel_settings/channel/list' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by allenlai18**
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-20580

https://github.com/mattermost/mattermost-server/issues/13492

### CODE REVIEW

**Comment by allenlai18** [comment]
@devinbinnie

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 1]
```diff
@@ -15,25 +15,36 @@ import SearchIcon from 'components/widgets/icons/search_icon';
...
-    static propTypes = {
-        actions: PropTypes.shape({
-            searchAllChannels: PropTypes.func.isRequired,
-            getData: PropTypes.func.isRequired,
-        }).isRequired,
-        data: PropTypes.array,
-        total: PropTypes.number,
+interface ChannelListProps {
+    actions: {
+        searchAllChannels: (term: string, notAssociatedToGroup?: string, excludeDefaultChannels?: boolean, page?: number, perPage?: number) => any;
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: can we avoid using 'any' here and type the response properly?

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 2]
```diff
@@ -15,25 +15,36 @@ import SearchIcon from 'components/widgets/icons/search_icon';
...
-        this.state = {
-            searchString: '',
-            channels: [],
-            searchTotalCount: 0,
-            pageResetKey: 0,
-        };
-    }
+interface ChannelListState {
+    searchString: string;
+    channels: undefined[];
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: undefined?

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 3]
```diff
@@ -60,41 +71,38 @@ export default class ChannelList extends React.PureComponent {
...
 
-    resetSearch = () => {
+    private resetSearch = () => {
         this.setState({searchString: '', channels: [], searchMode: false, searchTotalCount: 0, pageResetKey: Date.now()});
-    }
+    };
 
     header() {
         return (
-            <>
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: why the change? these are equivalent

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 4]
```diff
@@ -119,17 +127,17 @@ export default class ChannelList extends React.PureComponent {
                         <div className='group-actions'/>
                     </div>
                 </div>
-            </>
+            </React.Fragment>
         );
     }
 
-    onPageChangedCallback = (pagination, channels) => {
+    onPageChangedCallback = (pagination: {startCount: number; endCount: number; total: number}, channels: any[]) => {
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: can we avoid 'any' here?

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 5]
```diff
@@ -143,21 +151,21 @@ export default class ChannelList extends React.PureComponent {
                 onPageChangedCallback={this.onPageChangedCallback}
                 data={this.state.searchMode ? this.state.channels : this.props.data}
                 total={this.state.searchMode ? this.state.searchTotalCount : this.props.total}
-            />);
+            />
+        );
     }
 
-    renderRow = (item) => {
+    private renderRow = (item: any) => {
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: 'any' again

File: components/admin_console/team_channel_settings/channel/list/channel_row.tsx [hunk 6]
```diff
@@ -2,42 +2,44 @@
...
 import GlobeIcon from 'components/widgets/icons/globe_icon';
 import LockIcon from 'components/widgets/icons/lock_icon';
 
-export default class ChannelRow extends React.Component {
-    static propTypes = {
-        channel: PropTypes.object.isRequired,
-        onRowClick: PropTypes.func.isRequired,
-    };
+interface Props {
+    channel: any;
```
Comment (reflog) [inline components/admin_console/team_channel_settings/channel/list/channel_row.tsx]: 'any'

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx (hunk 2 above)
Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: @reflog  for channels. right now the channel and channel[] type is type any. Should I create a Ichannel interface in a separate file and use that? Should I take it all the fields from mm-server channel modal?

Comment (reflog) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: there's Channel definition in mattermost-redux and it should be imported from there

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 7]
```diff
@@ -15,25 +17,36 @@ import SearchIcon from 'components/widgets/icons/search_icon';
...
-            searchAllChannels: PropTypes.func.isRequired,
-            getData: PropTypes.func.isRequired,
-        }).isRequired,
-        data: PropTypes.array,
-        total: PropTypes.number,
+interface ChannelListProps {
+    actions: {
+        searchAllChannels: (term: string, notAssociatedToGroup?: string, excludeDefaultChannels?: boolean, page?: number, perPage?: number) => ActionFunc | ActionResult;
+        getData: (page: number, perPage: number, notAssociatedToGroup? : string, excludeDefaultChannels?: boolean) => ActionFunc | ActionResult;
+        removeGroup?: () => void;
```
Comment (devinbinnie) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: I don't think you need this Prop, it's not passed in through Redux

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 8]
```diff
@@ -15,25 +17,36 @@ import SearchIcon from 'components/widgets/icons/search_icon';
...
     };
+    data?: {id: string; display_name: string}[];
+    total?: number;
+    removeGroup?: () => void;
+    onPageChangedCallback?: () => void;
+    emptyListTextId: string;
```
Comment (devinbinnie) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: About these props used by tests, I don't even think the tests are using them. We can probably remove them from here, and the tests.

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 9]
```diff
@@ -15,25 +17,36 @@ import SearchIcon from 'components/widgets/icons/search_icon';
...
+interface ChannelListState {
+    searchString: string;
+    channels: Channel[];
+    searchTotalCount: number;
+    pageResetKey: number;
+    searchMode: boolean;
+}
+
+export default class ChannelList extends React.PureComponent<ChannelListProps, ChannelListState> {
+    state: ChannelListState = {
```
Comment (devinbinnie) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: These should be set using the `constructor()`

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx (hunk 8 above)
Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: you are right

Comment (devinbinnie) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: Nevermind, they're passed down into `AbstractList`, my bad. Nothing to be done here.

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx [hunk 10]
```diff
@@ -60,37 +73,38 @@ export default class ChannelList extends React.PureComponent {
...
         }
-
         if (searchString.length === 0) {
             this.resetSearch();
         }
-    }
-
-    getDataBySearch = async (page, perPage) => {
+    };
+    private getDataBySearch = async (page: number, perPage: number) => {
```
Comment (allenlai18) [inline components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: @reflog  How would you convert this async function to TS when it has to be assigned to ```absProps.actions.getData```   defined at line 23?

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx (hunk 7 above)
Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: > Thanks @allenlai18, this is looking great!
> As part of this ticket, can you also convert the redux connector to TypeScript as well?

you mean index.js right?  yes i will do that

**Comment by allenlai18** [comment]
recently updated index.js to TS as well

File: components/admin_console/team_channel_settings/channel/list/channel_list.tsx (hunk 10 above)
Comment (reflog) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: @allenlai18 - I've committed a possible fix. Let me know if this helps

Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: > @allenlai18 - I've committed a possible fix. Let me know if this helps

@reflog   This is a good solution and it simplifies it.  However, in index.ts mapDispatchToProps function wants the function interface to be the exact same from redux. How should I fix this?

Comment (reflog) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: @allenlai18  committed suggested fix

Comment (allenlai18) [reply components/admin_console/team_channel_settings/channel/list/channel_list.tsx]: ah yes. that is a way of doing it.

**Comment by allenlai18** [comment]
@reflog @devinbinnie   ready for review.    thank you @reflog  for all the help

**Comment by allenlai18** [comment]
@devinbinnie   can we reassign the QA?

**Comment by devinbinnie** [comment]
@saturninoabril do you have time to review this? If not, I can reassign.
